Sources/SPMLLBuild/llbuild.swift

```diff
@@ -8,10 +8,138 @@
  See http://swift.org/CONTRIBUTORS.txt for Swift project authors
 */
 
+import class Foundation.JSONEncoder
+import class Foundation.JSONDecoder
+import struct Foundation.Data
+
 #if canImport(llbuildSwift)
 @_exported import llbuildSwift
 #elseif canImport(llbuild)
 @_exported import llbuild
 #else
 // This should be a hard error in future.
 #endif
+
+public protocol LLBuildKey: Codable {
+}
+
+public protocol LLBuildValue: Codable {
+}
+
+public protocol LLBuildEngineDelegate {
+    func lookupRule(rule: String, key: Key) -> Rule
+}
+
+public final class LLBuildEngine {
+
+    private final class Delegate: BuildEngineDelegate {
+        let delegate: LLBuildEngineDelegate
+
+        init(_ delegate: LLBuildEngineDelegate) {
+            self.delegate = delegate
+        }
+
+        func lookupRule(_ key: Key) -> Rule {
+            let ruleKey = RuleKey(key)
+            return delegate.lookupRule(
+                rule: ruleKey.rule, key: Key(ruleKey.data))
+        }
+    }
+
+    private let engine: BuildEngine
+
+    public init(delegate: LLBuildEngineDelegate) {
+        engine = BuildEngine(delegate: Delegate(delegate))
+    }
+
+    public func build(key: Key) -> Value {
+        return engine.build(key: key)
+    }
+
+    public func attachDB(path: String, schemaVersion: Int = 0) throws {
+        try engine.attachDB(path: path, schemaVersion: schemaVersion)
+    }
+
+    public func close() {
+        engine.close()
+    }
+}
+
+/// An individual build rule.
+open class BuildRule<BuildKey: LLBuildKey>: Rule, Task {
+
+    /// The name of the rule.
+    ///
+    /// This name will be available in the delegate's lookupRule(rule:key:).
+    open class var ruleName: String {
+        fatalError("subclass responsibility")
+    }
+
+    public init() {
+    }
+
+    public func createTask() -> Task {
+        return self
+    }
+
+    open func start(_ engine: TaskBuildEngine) {
+    }
+
+    open func provideValue(_ engine: TaskBuildEngine, inputID: Int, value: Value) {
+    }
+
+    open func inputsAvailable(_ engine: TaskBuildEngine) {
+    }
+
+    public static func createKey(_ key: BuildKey) -> Key {
+        return RuleKey(
+            rule: self.ruleName, data: key.toKey().data).toKey()
+    }
+}
+
+// MARK:- Helpers
+
+public extension LLBuildKey {
+    init(_ key: Key) {
+        self.init(key.data)
+    }
+
+    init(_ data: [UInt8]) {
+        self = try! fromBytes(data)
```

Comment:
Needs to be prepared for errors

Reply:
Isn't it a programmer error if this fails?